Modify the kubeadm upgrade DAG for the TLS Upgrade

stealthybox opened the pull request:
**What this PR does / why we need it**:
This adds the necessary utilities to detect Etcd TLS on static pods from the file system and query Etcd.
It modifies the upgrade logic to make it support the APIServer downtime.
Tests are included and should be passing.

```bash 
bazel test //cmd/kubeadm/... \
  && bazel build //cmd/kubeadm --platforms=@io_bazel_rules_go//go/toolchain:linux_amd64 \
  && issue=TLSUpgrade ~/Repos/vagrant-kubeadm-testing/copy_kubeadm_bin.sh
```
These cases are working consistently for me
```bash
kubeadm-1.9.6 reset \
  && kubeadm-1.9.6 init --kubernetes-version 1.9.1 \
  && kubectl apply -f https://git.io/weave-kube-1.6
/vagrant/bin/TLSUpgrade_kubeadm upgrade apply 1.9.6  # non-TLS to TLS
/vagrant/bin/TLSUpgrade_kubeadm upgrade apply 1.10.0 # TLS to TLS
/vagrant/bin/TLSUpgrade_kubeadm upgrade apply 1.10.1 # TLS to TLS
/vagrant/bin/TLSUpgrade_kubeadm upgrade apply 1.9.1  # TLS to TLS /w major version downgrade
```

This branch is based on top of #61942, as resolving the hash race condition is necessary for consistent behavior.
It looks to fit in pretty well with @craigtracey's PR: #62141
The interfaces are pretty similar

/assign @detiber @timothysc

**Which issue(s) this PR fixes**
Helps with https://github.com/kubernetes/kubeadm/issues/740

**Special notes for your reviewer**:

278b322a1c
   [kubeadm] Implement ReadStaticPodFromDisk

c74b56372d
   Implement etcdutils with Cluster.HasTLS()

   - Test HasTLS()
   - Instrument throughout upgrade plan and apply
   - Update plan_test and apply_test to use new fake Cluster interfaces
   - Add descriptions to upgrade range test
   - Support KubernetesDir and EtcdDataDir in upgrade tests
   - Cover etcdUpgrade in upgrade tests
   - Cover upcoming TLSUpgrade in upgrade tests

8d8e5fe33b
   Update test-case, fix nil-pointer bug, and improve error message

97117fa873
   Modify the kubeadm upgrade DAG for the TLS Upgrade

   - Calculate `beforePodHashMap` before the etcd upgrade in anticipation of
   KubeAPIServer downtime
   - Detect if pre-upgrade etcd static pod cluster `HasTLS()==false` to switch
   on the Etcd TLS Upgrade if TLS Upgrade:
      - Skip L7 Etcd check (could implement a waiter for this)
      - Skip data rollback on etcd upgrade failure due to lack of L7 check
    (APIServer is already down unable to serve new requests)
      - On APIServer upgrade failure, also rollback the etcd manifest to
    maintain protocol compatibility

   - Add logging

**Release note**:
```release-note
kubeadm upgrade no longer races leading to unexpected upgrade behavior on pod restarts
kubeadm upgrade now successfully upgrades etcd and the controlplane to use TLS
kubeadm upgrade now supports external etcd setups
kubeadm upgrade can now rollback and restore etcd after an upgrade failure
```

detiber committed:
[kubeadm] fix mirror-pod hash race condition

- Update kubeadm static pod upgrades to use the
  kubetypes.ConfigHashAnnotationKey annotation on the mirror pod rather
  than generating a hash from the full object info. Previously, a status
  update for the pod would allow the upgrade to proceed before the
  new static pod manifest was actually deployed.

Signed-off-by: Jason DeTiberus <detiber@gmail.com>

stealthybox committed:
[kubeadm] Implement ReadStaticPodFromDisk

stealthybox committed:
[kubeadm] Implement etcdutils with Cluster.HasTLS()

- Test HasTLS()
- Instrument throughout upgrade plan and apply
- Update plan_test and apply_test to use new fake Cluster interfaces
- Add descriptions to upgrade range test
- Support KubernetesDir and EtcdDataDir in upgrade tests
- Cover etcdUpgrade in upgrade tests
- Cover upcoming TLSUpgrade in upgrade tests

stealthybox committed:
[kubeadm] Update test-case, fix nil-pointer bug, and improve error message

stealthybox committed:
[kubeadm] Modify the kubeadm upgrade DAG for the TLS Upgrade

- Calculate `beforePodHashMap` before the etcd upgrade in anticipation of KubeAPIServer downtime
- Detect if pre-upgrade etcd static pod cluster `HasTLS()==false` to switch on the Etcd TLS Upgrade
if TLS Upgrade:
  - Skip L7 Etcd check (could implement a waiter for this)
  - Skip data rollback on etcd upgrade failure due to lack of L7 check (APIServer is already down unable to serve new requests)
  - On APIServer upgrade failure, also rollback the etcd manifest to maintain protocol compatibility

- Add logging

timothysc commented on cmd/kubeadm/app/phases/upgrade/staticpods.go:
```diff
@@ -180,34 +181,56 @@ func upgradeComponent(component string, waiter apiclient.Waiter, pathMgr StaticP
+		// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+		// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+		// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+		if err := waiter.WaitForStaticPodHashChange(cfg.NodeName, component, beforePodHash); err != nil {
```
Doesn't this overlap with @detiber 's proposed kubelet change?

stealthybox replied:
It does, these are all proposed solutions that complement each other and overlap in different ways.

This PR skips the dependency on the APIServer for the protocol switch.

The kubelet API PR completely gets rid of the hard dependency on the APIServer.

timothysc reviewed (changes requested):
I'm going to stop reviews on this until we can chat.  

However, I would like to trim out any refactorings that may be unnecessary in this PR.

stealthybox commented on cmd/kubeadm/app/phases/upgrade/compute.go:
```diff
@@ -23,7 +23,7 @@ import (
 	"k8s.io/kubernetes/cmd/kubeadm/app/features"
 	"k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns"
-	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+	etcdutil "k8s.io/kubernetes/cmd/kubeadm/app/util/etcd"
```
I neglected to put this in the commit, but it was necessary for me to restructure the module, because `HasTLS()` imports `ReadStaticPodFromDisk ` from `app/util/staticpod`

`app/util/staticpod` already imports from `/app/util` and this created a dependency cycle.

detiber committed:
[kubeadm] Add etcd L7 check on upgrade

- Adds L7 check for kubeadm etcd static pod upgrade

stealthybox committed:
[kubeadm] Fix Etcd Rollback

Fix `rollbackEtcdData()` to return error=nil on success
`rollbackEtcdData()` used to always return an error making the rest of the
upgrade code completely unreachable.

Ignore errors from `rollbackOldManifests()` during the rollback since it
always returns an error.
Success of the rollback is gated with etcd L7 healthchecks.

Remove logic implying the etcd manifest should be rolled back when
`upgradeComponent()` fails

timothysc commented:
@stealthybox @detiber - we need to get this done ASAP for 1.10.2 release!

stealthybox commented:
/retest

stealthybox commented:
@timothysc 
I think @detiber and I agree this patch is near complete.
We've validated the happy path and improved some error cases.

We should validate that it also allows for External Etcd during the upgrade such as in #62141 
(cc @craigtracey)

A final nice to have would be to add more test cases and reduce the differences between test behavior and actual behavior.

detiber commented:
@stealthybox There are still some additional changes needed to unblock external etcd, but I think I'd like to get that sorted out as a followup.

detiber commented:
/lgtm

timothysc commented:
/cc @MaciekPytel - for milestone cherry-pick approval.

timothysc commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -281,7 +282,9 @@ func PerformStaticPodUpgrade(client clientset.Interface, waiter apiclient.Waiter
 
-	return upgrade.StaticPodControlPlane(waiter, pathManager, internalcfg, etcdUpgrade)
+	// These are uninitialized because passing in the clients allow for mocking the client during testing
+	var oldEtcdClient, newEtdClient etcdutil.Client
```
This is weird and would cause static analysis issues.

stealthybox replied:
We could pass the client creation code as functions?
WDYT?

timothysc commented on cmd/kubeadm/app/phases/upgrade/compute_test.go:
```diff
@@ -61,14 +62,20 @@ func (f *fakeVersionGetter) KubeletVersions() (map[string]uint16, error) {
 }
 
-type fakeEtcdCluster struct{}
+type fakeEtcdCluster struct{ TLS bool }
```
Could you please open an issue about instead of creating a stub to use the test jiggery that exists to auto-standup and cleanup etcd.

timothysc commented on cmd/kubeadm/app/phases/upgrade/staticpods_test.go:
```diff
@@ -193,14 +209,60 @@ func (spm *fakeStaticPodPathManager) BackupEtcdDir() string {
+	return c.TLS
+}
+
+func (c fakeTLSEtcdClient) GetStatus() (*clientv3.StatusResponse, error) {
```
same as other comment, I'm an anti-stub person and we have the jiggery to dynamically spin clusters in UTs.

stealthybox replied:
@timothysc this sounds great for testing `GenericClient.GetStatus()` in `util/etcd/etcd_test.go`

Do we have an existing unit-test harness for running etcd as a static pod using a kubelet?
That would be what is required for this particular test.

timothysc replied:
We have utilities to basically spin up an actual etcd server + client pair, but at this point I'm ok with merging what is here and we can create a new issue and cross-link for future work.

timothysc commented on cmd/kubeadm/app/phases/upgrade/staticpods_test.go:
```diff
@@ -136,29 +141,36 @@ type fakeStaticPodPathManager struct {
 
 func NewFakeStaticPodPathManager(moveFileFunc func(string, string) error) (StaticPodPathManager, error) {
-	realManifestsDir, err := ioutil.TempDir("", "kubeadm-upgraded-manifests")
+	kubernetesDir, err := ioutil.TempDir("", "kubeadm-pathmanager-")
```
Who cleans up these directories?

stealthybox replied:
https://github.com/kubernetes/kubernetes/pull/62655/files#diff-012a4c853bcc083f2ac6c61d69e7292eR375

timothysc commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+// GenericClient is a common etcd client for supported etcd servers
+type GenericClient struct {
```
Did you test the upgrade scenario of external etcd that @craigtracey was validating?

stealthybox replied:
We don't have a test for this, but I assume @detiber would include it in the follow up PR?

timothysc commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -0,0 +1,168 @@
+func (c GenericClient) GetStatus() (*clientv3.StatusResponse, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   c.Endpoints,
+		DialTimeout: 5 * time.Second,
```
const above.

timothysc commented on cmd/kubeadm/app/util/etcd/etcd.go:
```diff
@@ -0,0 +1,168 @@
+// WaitForStatus returns a StatusResponse after an initial delay and retry attempts
+func (c GenericClient) WaitForStatus(delay time.Duration, retries int, retryInterval time.Duration) (*clientv3.StatusResponse, error) {
+	fmt.Printf("[util/etcd] Waiting %v for initial delay\n", delay)
+	time.Sleep(delay)
```
this is basically a wait.Until, please use that f(n) or comment to reduce this later.

stealthybox replied:
@timothysc, are you referring to this?
https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Until

detiber replied:
https://github.com/stealthybox/kubernetes/pull/2 attempts to modify WaitForStatus to use wait.Until from apimachinery.

timothysc commented on cmd/kubeadm/app/util/etcd/etcd_test.go:
```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
2018

timothysc reviewed (changes requested):
Please address comments. 

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews  

for more eyeballs, b/c this is a large back-patch and we want to be certain here.

stealthybox commented:
/retest

timothysc reviewed (approved):
/lgtm 
/approve

discostur commented:
Shouldn't this now also work for external etcd clusters? Because i just updated kubeadm to 1.10.2 but i still get:

``` 
$kubeadm upgrade plan
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
could not read manifests from: /etc/kubernetes/manifests, error: failed to check if etcd pod implements TLS: failed to read manifest for "/etc/kubernetes/manifests/etcd.yaml": open /etc/kubernetes/manifests/etcd.yaml: no such file or directory
```

ypsingh27 commented:
I have upgraded my kubeadm and kubectl version to v1.10.2 
root # kubeadm version
kubeadm version: &version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-04-27T09:10:24Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
root # kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-04-27T09:22:21Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.3", GitCommit:"d2835416544f298c919e2ead3be3d0864b52323b", GitTreeState:"clean", BuildDate:"2018-02-07T11:55:20Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}


I am still not able to upgrade successfully. I get the below error:

root # kubeadm upgrade apply v1.10.1
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade/version] You have chosen to change the cluster version to "v1.10.1"
[upgrade/versions] Cluster version: v1.9.3
[upgrade/versions] kubeadm version: v1.10.2
[upgrade/confirm] Are you sure you want to proceed with the upgrade? [y/N]: y
[upgrade/prepull] Will prepull images for components [kube-apiserver kube-controller-manager kube-scheduler]
[upgrade/apply] Upgrading your Static Pod-hosted control plane to version "v1.10.1"...
Static pod: kube-apiserver-vmaksa69901dzl hash: dc3749dffa0c124bd5c4964613658249
Static pod: kube-controller-manager-vmaksa69901dzl hash: 9e6798d0ba2ebe6747904b2195183c11
Static pod: kube-scheduler-vmaksa69901dzl hash: d38a1ee5cf80a84bc4f295d19b6874c2
[upgrade/etcd] Upgrading to TLS for etcd
Static pod: etcd-vmaksa69901dzl hash: 17c801f54fe8bd173a9f4810b4242bbb
[etcd] Wrote Static Pod manifest for a local etcd instance to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests729921876/etcd.yaml"
[certificates] Using the existing etcd/ca certificate and key.
[certificates] Using the existing etcd/server certificate and key.
[certificates] Using the existing etcd/peer certificate and key.
[certificates] Using the existing etcd/healthcheck-client certificate and key.
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/etcd.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests825459363/etcd.yaml"
[upgrade/staticpods] Not waiting for pod-hash change for component "etcd"
[upgrade/etcd] Waiting for etcd to become available
[util/etcd] Waiting 30s for initial delay
[util/etcd] Attempting to get etcd status 1/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: getsockopt: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 2/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: getsockopt: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 3/10
[util/etcd] Attempt failed with error: dial tcp [::1]:2379: getsockopt: connection refused
[util/etcd] Waiting 15s until next retry
[util/etcd] Attempting to get etcd status 4/10
[upgrade/staticpods] Writing new Static Pod manifests to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests729921876"
[controlplane] Wrote Static Pod manifest for component kube-apiserver to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests729921876/kube-apiserver.yaml"
[controlplane] Wrote Static Pod manifest for component kube-controller-manager to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests729921876/kube-controller-manager.yaml"
[controlplane] Wrote Static Pod manifest for component kube-scheduler to "/etc/kubernetes/tmp/kubeadm-upgraded-manifests729921876/kube-scheduler.yaml"
[upgrade/staticpods] The etcd manifest will be restored if component "kube-apiserver" fails to upgrade
[certificates] Using the existing etcd/ca certificate and key.
[certificates] Using the existing apiserver-etcd-client certificate and key.
[upgrade/staticpods] Moved new manifest to "/etc/kubernetes/manifests/kube-apiserver.yaml" and backed up old manifest to "/etc/kubernetes/tmp/kubeadm-backup-manifests825459363/kube-apiserver.yaml"
[upgrade/staticpods] Waiting for the kubelet to restart the component
[upgrade/apply] FATAL: couldn't upgrade control plane. kubeadm has tried to recover everything into the earlier state. Errors faced: [timed out waiting for the condition]

Can someone help me with this upgrade? any worarounds?

stealthybox commented:
@ypsingh27 it looks like your upgrade failed on the apiserver manifest.
It appears kubeadm rolled back your cluster to the previous functioning state. Is that correct?

The first thing I think of is that you may not have been able to pull the image within the timeout.
Pre-pulling images can help with the upgrade.
Beyond that, I can't help without more information about your config and environment.

Somebody in the [#kubeadm slack channel](https://kubernetes.slack.com/messages/C2P1JHS2E/) may be able to help you.

discostur commented:
@detiber are there any news on the failed upgrade with external etcd support? Because in the 1.10.2 release notes they write:

```
kubeadm upgrade now supports external etcd setups
``` 

but it still fails :/

robertrbruno commented:
Just wanted to second the same issues @discostur is seeing.   Any updates?

discostur commented:
@robertrbruno https://github.com/kubernetes/kubernetes/pull/63495
